thirdparty/github.go

```diff
@@ -32,54 +30,84 @@ const (
+	if _, ok := err.(*github.RateLimitError); ok {
+		return false, err
+	}
+	if resp.StatusCode == http.StatusBadRequest {
+		return false, err
+	}
+	if resp.StatusCode == http.StatusNotFound {
+		return false, err
+	}
+	if resp.StatusCode == http.StatusUnauthorized {
+		err = errors.Errorf("Calling github GET on %v failed: got 'unauthorized' response", resp.Request.URL.String())
+		grip.Error(err)
+		return false, err
+	}
+	if err != nil {
+		grip.Errorf("failed trying to call github GET on %s: %+v", resp.Request.URL.String(), err)
+		return true, err
+	}
+	if resp.StatusCode != http.StatusOK {
+		err = errors.Errorf("Calling github GET %s got a bad response code: %v", resp.Request.URL.String(), resp.StatusCode)
+	}
+
+	rateMessage, _ := getGithubRateLimit(resp.Header)
+	grip.Debugf("Github API response: %s. %s", resp.Status, rateMessage)
+
+	return false, err
 }
 
 // GetGithubCommits returns a slice of GithubCommit objects from
 // the given commitsURL when provided a valid oauth token
-func GetGithubCommits(oauthToken, commitsURL string) (
-	githubCommits []github.RepositoryCommit, header http.Header, err error) {
-	resp, err := tryGithubGet(oauthToken, commitsURL)
-	if resp != nil {
-		defer resp.Body.Close()
-	}
-	if resp == nil {
-		errMsg := fmt.Sprintf("nil response from url '%v'", commitsURL)
-		grip.Error(errMsg)
-		return nil, nil, APIResponseError{errMsg}
-	}
+func GetGithubCommits(oauthToken, owner, repo, ref string, commitPage int) ([]*github.RepositoryCommit, int, error) {
+	httpClient, err := util.GetHttpClientForOauth2(oauthToken)
 	if err != nil {
-		errMsg := fmt.Sprintf("error querying '%v': %v", commitsURL, err)
-		grip.Error(errMsg)
-		return nil, nil, APIResponseError{errMsg}
+		return nil, 0, errors.Wrap(err, "can't fetch data from github")
 	}
+	defer util.PutHttpClientForOauth2(httpClient)
+
+	client := github.NewClient(httpClient)
+	var commits []*github.RepositoryCommit
+	var resp *github.Response
+	_, _ = util.Retry(
+		func() (bool, error) {
+			commits, resp, err = client.Repositories.ListCommits(context.TODO(), owner, repo, &github.CommitsListOptions{
+				SHA: ref,
+				ListOptions: github.ListOptions{
+					Page: commitPage,
+				},
+			})
+
+			return githubRetry(resp, err)
+		}, NumGithubRetries, GithubSleepTimeSecs*time.Second)
 
-	header = resp.Header
-	respBody, err := ioutil.ReadAll(resp.Body)
 	if err != nil {
-		return nil, nil, ResponseReadError{err.Error()}
+		errMsg := fmt.Sprintf("error querying for commits in '%s/%s' ref %s : %v", owner, repo, ref, err)
+		grip.Error(errMsg)
+		return nil, 0, APIResponseError{errMsg}
 	}
-
-	grip.Debugf("Github API response: %s. %d bytes", resp.Status, len(respBody))
+	if resp == nil {
+		errMsg := fmt.Sprintf("nil response from url '%s/%s' ref %s", owner, repo, ref)
+		grip.Error(errMsg)
+		return nil, 0, APIResponseError{errMsg}
+	}
+	defer resp.Body.Close()
 
 	if resp.StatusCode != http.StatusOK {
```

Comment:
Can we ever actually get here? The retry func checks the status code and sets err, which we then check on 86.